File: python/peer.py
#!/usr/bin/env python3
"""
secure-p2p — Encrypted P2P File Sharing CLI Tool
Stage 3: AES-256-CBC encryption wired into file transfer

Usage:
    Receive mode: python3 peer.py receive [--port PORT]
    Send mode:    python3 peer.py send <filepath> <host> [--port PORT]
"""

import socket
import struct
import os
import sys
import argparse
from crypto_utils import (
    perform_handshake_sender,
    perform_handshake_receiver,
    aes_encrypt,
    aes_decrypt,
)


# ─── Constants ────────────────────────────────────────────────────────────────

DEFAULT_PORT  = 9999
CHUNK_SIZE    = 65536   # 64 KB per chunk
HEADER_FORMAT = ">Q"    # Big-endian unsigned 64-bit integer (8 bytes)
HEADER_SIZE   = struct.calcsize(HEADER_FORMAT)


# ─── Low-level send / recv helpers ────────────────────────────────────────────

def send_bytes(sock: socket.socket, data: bytes) -> None:
    """Send length-prefixed bytes. Protocol: [8-byte length][data]"""
    header = struct.pack(HEADER_FORMAT, len(data))
    sock.sendall(header)
    sock.sendall(data)


def recv_bytes(sock: socket.socket) -> bytes:
    """Receive one complete length-prefixed message."""
    header = _recv_exact(sock, HEADER_SIZE)
    if not header:
        raise ConnectionError("Connection closed before header received")
    (length,) = struct.unpack(HEADER_FORMAT, header)
    data = _recv_exact(sock, length)
    if not data:
        raise ConnectionError("Connection closed before full message received")
    return data


def _recv_exact(sock: socket.socket, n: int) -> bytes:
    """Read exactly n bytes, looping until all arrive."""
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("Socket closed mid-stream")
        buf.extend(chunk)
    return bytes(buf)


# ─── File transfer (AES-256-CBC encrypted) ────────────────────────────────────

def send_file(sock: socket.socket, filepath: str, aes_key: bytes) -> None:
    """
    Send a file over an established socket connection, encrypted with AES-256-CBC.

    What gets encrypted:
      - The filename (so even the filename is private)
      - Every chunk of file data

    What stays plaintext:
      - The file size (receiver needs this to know when transfer is complete)
        In a production system you'd encrypt this too and use authenticated
        encryption (AES-GCM) — noted in README security considerations.
    """
    filename = os.path.basename(filepath)
    filesize = os.path.getsize(filepath)

    print(f"[*] Sending '{filename}' ({filesize:,} bytes) — AES-256-CBC encrypted")

    # Encrypt and send filename
    send_bytes(sock, aes_encrypt(filename.encode("utf-8"), aes_key))

    # Send file size plaintext (needed for progress tracking)
    send_bytes(sock, struct.pack(HEADER_FORMAT, filesize))

    # Encrypt and send file data chunk by chunk
    bytes_sent = 0
    with open(filepath, "rb") as f:
        while True:
            chunk = f.read(CHUNK_SIZE)
            if not chunk:
                break
            encrypted_chunk = aes_encrypt(chunk, aes_key)
            send_bytes(sock, encrypted_chunk)
            bytes_sent += len(chunk)
            print(f"\r[*] Progress: {bytes_sent:,} / {filesize:,} bytes", end="", flush=True)

    print(f"\n[+] File sent and encrypted successfully: {bytes_sent:,} bytes")


def receive_file(sock: socket.socket, aes_key: bytes, save_dir: str = ".") -> str:
    """
    Receive and decrypt a file from an established socket connection.
    """
    # Decrypt filename
    filename = aes_decrypt(recv_bytes(sock), aes_key).decode("utf-8")

    # Receive plaintext file size
    (filesize,) = struct.unpack(HEADER_FORMAT, recv_bytes(sock))

    print(f"[*] Receiving '{filename}' ({filesize:,} bytes) — decrypting AES-256-CBC")

    save_path      = os.path.join(save_dir, f"received_{filename}")
    bytes_received = 0

    with open(save_path, "wb") as f:
        while bytes_received < filesize:
            encrypted_chunk = recv_bytes(sock)
            chunk = aes_decrypt(encrypted_chunk, aes_key)
            f.write(chunk)
            bytes_received += len(chunk)
            print(f"\r[*] Progress: {bytes_received:,} / {filesize:,} bytes", end="", flush=True)

    print(f"\n[+] File received and decrypted: {save_path}")
    return save_path


# ─── Receive mode ─────────────────────────────────────────────────────────────

def run_receiver(port: int) -> None:
    server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server_sock.bind(("0.0.0.0", port))
    server_sock.listen(1)

    print(f"[*] Listening on 0.0.0.0:{port} — waiting for sender...")

    try:
        conn, addr = server_sock.accept()
        print(f"[+] Connection established from {addr[0]}:{addr[1]}")

        with conn:
            aes_key = perform_handshake_receiver(conn)
            receive_file(conn, aes_key)

    except KeyboardInterrupt:
        print("\n[!] Receiver stopped by user")
    finally:
        server_sock.close()


# ─── Send mode ────────────────────────────────────────────────────────────────

def run_sender(filepath: str, host: str, port: int) -> None:
    if not os.path.exists(filepath):
        print(f"[!] Error: file not found: {filepath}")
        sys.exit(1)

    print(f"[*] Connecting to {host}:{port} ...")

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    try:
        sock.connect((host, port))
        print(f"[+] Connected to {host}:{port}")

        with sock:
            aes_key = perform_handshake_sender(sock)
            send_file(sock, filepath, aes_key)

    except ConnectionRefusedError:
        print(f"[!] Connection refused — is the receiver running on {host}:{port}?")
        sys.exit(1)
    except Exception as e:
        print(f"[!] Error: {e}")
        sys.exit(1)


# ─── CLI entry point ──────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(
        description="secure-p2p: Encrypted P2P file transfer",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  Start receiver:      python3 peer.py receive
  Start receiver:      python3 peer.py receive --port 8888
  Send a file:         python3 peer.py send photo.jpg 192.168.1.5
  Send a file:         python3 peer.py send photo.jpg 192.168.1.5 --port 8888
        """
    )

    subparsers = parser.add_subparsers(dest="mode", required=True)

    recv_parser = subparsers.add_parser("receive", help="Listen for incoming file transfer")
    recv_parser.add_argument("--port", type=int, default=DEFAULT_PORT,
                             help=f"Port to listen on (default: {DEFAULT_PORT})")

    send_parser = subparsers.add_parser("send", help="Send a file to a receiver")
    send_parser.add_argument("filepath", help="Path to the file to send")
    send_parser.add_argument("host",     help="Receiver's IP address or hostname")
    send_parser.add_argument("--port",   type=int, default=DEFAULT_PORT,
                             help=f"Receiver's port (default: {DEFAULT_PORT})")

    args = parser.parse_args()

    if args.mode == "receive":
        run_receiver(args.port)
    elif args.mode == "send":
        run_sender(args.filepath, args.host, args.port)


if __name__ == "__main__":
    main()

File: c/peer.c
#include "peer.h"
#include "crypto_utils.h"
#include "transfer.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <unistd.h>

#ifdef __APPLE__
#include <libkern/OSByteOrder.h>
#define htobe64(x) OSSwapHostToBigInt64(x)
#define be64toh(x) OSSwapBigToHostInt64(x)
#endif

/* ── Low-level socket helpers ─────────────────────────────────────────────── */

int send_bytes(int sock, const uint8_t *data, size_t len) {
    /*
     * Send a length-prefixed message.
     * Protocol: [8-byte big-endian uint64 length][data]
     *
     * send() may transmit fewer bytes than requested if the kernel
     * send buffer is full — we loop until all bytes are sent.
     */
    uint64_t len_be = htobe64((uint64_t)len);
    uint8_t  header[HEADER_SIZE];
    memcpy(header, &len_be, HEADER_SIZE);

    /* Send header */
    size_t sent = 0;
    while (sent < HEADER_SIZE) {
        ssize_t s = send(sock, header + sent, HEADER_SIZE - sent, 0);
        if (s <= 0) return -1;
        sent += (size_t)s;
    }

    /* Send data */
    sent = 0;
    while (sent < len) {
        ssize_t s = send(sock, data + sent, len - sent, 0);
        if (s <= 0) return -1;
        sent += (size_t)s;
    }

    return 0;
}

int recv_exact(int sock, uint8_t *buf, size_t n) {
    /*
     * Read exactly n bytes from sock.
     * TCP may deliver data in smaller chunks than requested — we loop.
     */
    size_t received = 0;
    while (received < n) {
        ssize_t r = recv(sock, buf + received, n - received, 0);
        if (r <= 0) return -1;
        received += (size_t)r;
    }
    return 0;
}

ssize_t recv_bytes(int sock, uint8_t **out_buf) {
    /*
     * Receive one complete length-prefixed message.
     * Allocates a heap buffer — caller must free(*out_buf).
     */
    uint8_t header[HEADER_SIZE];
    if (recv_exact(sock, header, HEADER_SIZE) != 0) return -1;

    uint64_t len_be;
    memcpy(&len_be, header, HEADER_SIZE);
    size_t len = (size_t)be64toh(len_be);

    uint8_t *buf = malloc(len);
    if (!buf) return -1;

    if (recv_exact(sock, buf, len) != 0) {
        free(buf);
        return -1;
    }

    *out_buf = buf;
    return (ssize_t)len;
}


/* ── Receive mode ─────────────────────────────────────────────────────────── */

static void run_receiver(int port) {
    int server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd < 0) { perror("[!] socket"); return; }

    /* SO_REUSEADDR — lets us restart immediately without 'address in use' */
    int opt = 1;
    setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    struct sockaddr_in addr = {0};
    addr.sin_family      = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port        = htons((uint16_t)port);

    if (bind(server_fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("[!] bind"); close(server_fd); return;
    }

    listen(server_fd, 1);
    printf("[*] Listening on 0.0.0.0:%d — waiting for sender...\n", port);

    struct sockaddr_in client_addr;
    socklen_t client_len = sizeof(client_addr);
    int conn = accept(server_fd, (struct sockaddr *)&client_addr, &client_len);
    if (conn < 0) { perror("[!] accept"); close(server_fd); return; }

    char client_ip[INET_ADDRSTRLEN];
    inet_ntop(AF_INET, &client_addr.sin_addr, client_ip, sizeof(client_ip));
    printf("[+] Connection established from %s:%d\n",
           client_ip, ntohs(client_addr.sin_port));

    uint8_t aes_key[AES_KEY_BYTES];
    if (perform_handshake_receiver(conn, aes_key) != 0) {
        fprintf(stderr, "[!] Handshake failed\n");
        close(conn); close(server_fd); return;
    }

    receive_file(conn, aes_key, ".");

    close(conn);
    close(server_fd);
}


/* ── Send mode ────────────────────────────────────────────────────────────── */

static void run_sender(const char *filepath, const char *host, int port) {
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) { perror("[!] socket"); return; }

    /* Resolve hostname to IP */
    struct hostent *he = gethostbyname(host);
    if (!he) {
        fprintf(stderr, "[!] Cannot resolve host: %s\n", host);
        close(sock); return;
    }

    struct sockaddr_in server_addr = {0};
    server_addr.sin_family = AF_INET;
    server_addr.sin_port   = htons((uint16_t)port);
    memcpy(&server_addr.sin_addr, he->h_addr_list[0], (size_t)he->h_length);

    printf("[*] Connecting to %s:%d ...\n", host, port);

    if (connect(sock, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        perror("[!] connect");
        fprintf(stderr, "[!] Is the receiver running on %s:%d?\n", host, port);
        close(sock); return;
    }

    printf("[+] Connected to %s:%d\n", host, port);

    uint8_t aes_key[AES_KEY_BYTES];
    if (perform_handshake_sender(sock, aes_key) != 0) {
        fprintf(stderr, "[!] Handshake failed\n");
        close(sock); return;
    }

    send_file(sock, filepath, aes_key);
    close(sock);
}


/* ── CLI entry point ──────────────────────────────────────────────────────── */

static void print_usage(const char *prog) {
    printf("Usage:\n");
    printf("  Receive: %s receive [--port PORT]\n", prog);
    printf("  Send:    %s send <filepath> <host> [--port PORT]\n\n", prog);
    printf("Examples:\n");
    printf("  %s receive\n", prog);
    printf("  %s receive --port 8888\n", prog);
    printf("  %s send photo.jpg 192.168.1.5\n", prog);
    printf("  %s send photo.jpg 192.168.1.5 --port 8888\n", prog);
}

int main(int argc, char *argv[]) {
    if (argc < 2) { print_usage(argv[0]); return 1; }

    int port = DEFAULT_PORT;

    if (strcmp(argv[1], "receive") == 0) {
        /* Parse optional --port */
        for (int i = 2; i < argc - 1; i++) {
            if (strcmp(argv[i], "--port") == 0) {
                port = atoi(argv[i + 1]);
            }
        }
        run_receiver(port);

    } else if (strcmp(argv[1], "send") == 0) {
        if (argc < 4) { print_usage(argv[0]); return 1; }
        const char *filepath = argv[2];
        const char *host     = argv[3];
        for (int i = 4; i < argc - 1; i++) {
            if (strcmp(argv[i], "--port") == 0) {
                port = atoi(argv[i + 1]);
            }
        }
        run_sender(filepath, host, port);

    } else {
        print_usage(argv[0]);
        return 1;
    }

    return 0;
}